File: DataStructure/list/reverselinkedlist.c
#include<stdio.h> 
#include<stdlib.h>

struct Node
{
	int data;
	struct Node * next;
};

/**
 *      c       
 *    +---+   +---+   +---+   +---+
 *    | 5 |-->| 4 |-->| 7 |-->| 9 | 
 *    +---+   +---+   +---+   +---+
 *    括号中的数字代表第几步
 *      c->n(2) n(1)
 *      p(3)    c(4)
 *    +---+   +---+   +---+   +---+
 *    | 5 |-->| 4 |-->| 7 |-->| 9 | 
 *    +---+   +---+   +---+   +---+
 */
static void reverse(struct Node ** head_ref)
{
	struct Node* prev = NULL;
	struct Node* current = *head_ref;
	struct Node* next = NULL;
	while(current != NULL) {
		next = current->next;
		current->next = prev;
		prev = current;
		current = next;
	}	
	*head_ref = prev;
}

void push(struct Node ** head_ref, int new_data)
{
	struct Node * new_node = (struct Node*)malloc(sizeof(struct Node));
	new_node->data = new_data;
	new_node->next = (*head_ref);
	(*head_ref) = new_node;
}

void printList(struct Node* head) 
{
	struct Node *tmp = head;
	while(tmp != NULL)
	{
		printf("%d ", tmp->data);
		tmp = tmp->next;
	}
}

int main(void)
{
	struct Node* head = NULL;
	push(&head, 20); 
	push(&head, 14); 
	push(&head, 5);  
	push(&head, 12); 
	
	printf("Given linked list\n"); 
	printList(head);     
	reverse(&head);                       
	printf("\nReversed Linked list \n"); 
	printList(head);     
	getchar(); 	
}

File: DataStructure/list/Reverselinkedlist.java
public Reverselinkedlist {
    static Node head; 
  
    static class Node { 
  
        int data; 
        Node next; 
  
        Node(int d) 
        { 
            data = d; 
            next = null; 
        } 
    } 
  
    /** 
    * Function to reverse the linked list
    * 
    * 单向链表
    */
    Node reverse(Node node) 
    { 
        Node prev = null;
        Node current = node;
        Node next = null;
        // 1. next前移 2. current.next翻转  3. pre前移  4. current前移
        // 结束条件是什么？ 一定是current会不断的后移
        while (current != null) {
            //向链表后移动，会到null
            next = current.next;
            //当前Node的后续node，指向原来的前向节点
            current.next = prev;
            // 注意这一步，避免前序node丢失
            prev = current;


            // 
            current = next;
        }
        node = prev;

        return node;
    }

    public void addFirst(int object) {
        Node newNode = new Node(object);
        newNode.next = head;
        head = newNode;
    }

    public void addLast(int object) { // here I don't need a tail reference
        Node temp = head;
        while(temp.next != null) {
            temp = temp.next;
        }

        temp.next = new Node(object);
    } 

    // prints content of double linked list 
    void printList(Node node) 
    { 
        while (node != null) { 
            System.out.print(node.data + " "); 
            node = node.next; 
        } 
    } 
  
    public static void main(String[] args) 
    { 
        LinkedList list = new LinkedList(); 
        list.head = new Node(85); 
        list.head.next = new Node(15); 
        list.head.next.next = new Node(4); 
        list.head.next.next.next = new Node(20); 
  
        System.out.println("Given Linked list"); 
        list.printList(head); 
        head = list.reverse(head); 
        System.out.println(""); 
        System.out.println("Reversed linked list "); 
        list.printList(head); 
    } 
}